api.yml

```diff
@@ -1,159 +1,62 @@
         content:
           application/json:
             schema:
-              $ref: '#/components/schemas/IssueRequest'
-        description: Parameters for issuing the credential.
+              $ref: '#/components/schemas/SignCredentialRequest'
+        description: Parameters for constructing and issuing the credential.
       responses:
         '201':
-          description: credential successfully issued!
+          description: Credential was successfully issued.
           content:
             application/json:
               schema:
-                type: object
-                description: The issued credential. This SHOULD include a reference or ID that can be used later for other parts of the API, e.g. changing and retrieving the status of the credential, or refreshing the credential.
+                $ref: '#/components/schemas/CredentialResponse'
         '400':
-          description: invalid input!
+          description: Malformed request from client.
         '500':
-          description: error!
-  /credential/{credentialReference}:
-    get:
-      tags:
-        - internal
-      summary: Retrieves a credential that has already been issued, and returns it in the response body.
-      operationId: retrieveCredential
-      description: Retrieves a credential that has already been issued, and returns it in the response body. The retrieved credential MUST be identical to the original credential when it was first issued. This can not be used for refreshing an expired credential. Support of this part of the API is OPTIONAL for implementations.
-      parameters:
-        - in: path
-          required: true
-          name: credentialReference
-          schema:
-            type: string
-          description: Reference to a credential that has been issued.
-      responses:
-        '200':
-          description: credential successfully retrieved!
-          content:
-            application/json:
-              schema:
-                type: object
-                description: The retrieved credential.
-        '400':
-          description: invalid input!
-        '404':
-          description: credential not found!
-        '500':
-          description: error!
-        '501':
-          description: not supported by this implementation!
-  /credential/{credentialReference}/status:
+          description: Internal error
+  /credentials/construct:
     post:
-      tags:
-        - internal
-      summary: Changes the credential status.
-      operationId: changeCredentialStatus
-      description: Changes the credential status. Support of this part of the API is OPTIONAL for implementations.
-      parameters:
-        - in: path
-          required: true
-          name: credentialReference
-          schema:
-            type: string
-          description: Reference to a credential that has been issued.
+      summary: Constructs a credential from pre-configured templates and returns it in the response body.
```

Comment:
> that's a more complex use case than the more immediate one in front of us (which is effectively just syntax checking, transforming, and signing).

I'm not sure where this use case is defined. Can you point me in the right direction?

> What is a template? My concern here is that if we add templates into the mix, then we have to agree on what the syntax for the template is going to be, which is going to be another long discussion that we don't have time for right now.

We think templates are a more sensible option in the long run by providing more value to clients and thus we'd like to start with that. For this first cut we can certainly define it in the simplest terms possible due to time constraints.

Reply:
In the current API definition I also felt that a templating feature could be useful:

```
      parameters:
        - in: query
          required: false
          name: templateReference
          schema:
            type: string
          description: Optional internal reference to a credential template. If this is present, it may be used to automatically pre-populate certain fields in the request body.
```

My idea was that this would be some internal reference and implementation-specific, and that the credential format doesn't necessarily have to be standardized (which I agree would be a lot of work).

Reply:
> this would be some internal reference and implementation-specific

Translation: "This would be some proprietary solution that wouldn't allow the customer to switch from one issuer platform to another."

:)

See the problem? This is a non-starter for a standard API because it'll inevitably lead to various vendors using different templating languages and thus different input parameters to the template. Again, we see the value in having a templating language, but if you don't standardize it, there is no interoperability... only vendor lock in.